docs/sql_reference/src/asciidoc/_chapters/sql_functions_and_expressions.adoc

```diff
@@ -487,6 +487,141 @@ LARGEINT if the precision of the argument is greater than or equal to
 ABS (-20 + 12)
 ```
 
+<<<
+[[aes_encrypt_function]]
+== AES_ENCRYPT Function
+ 
+The AES_ENCRYPT function encrypts a `_string_` using a specified encryption `_key_` with the AES (Advanced Encryption Standard) algorithm and returns a binary string. 
+
+This function prevents sensitive data from being visible to the public. For example, use the AES_ENCRYPT function to encrypt person identity number to preserve user confidentiality.
+
+`AES_ENCRYPT (_string_, _key_)`
+
+[[syntax_description_of_aes_encrypt]]
+=== Syntax Description of AES_ENCRYPT
+
+* `_string_` 
++
+is the data to be encrypted. 
+
+* `_key_` 
++
+is the encryption key to encrypt the `_string_`.
+
++
+TIP: The CQD `block_encryption_mode` controls the mode for the block-based encryption algorithm. The default mode is the aes-128-ecb, which means the encryption using a key length of 128 bits and the ECB mode.
+
++
+[cols="50%,50%"]
+|===
+^| *Value* ^| *Mode*
+^| 0 ^| aes-128-ecb
+^| 1 ^| aes-192-ecb
+^| 2 ^| aes-256-ecb
+|===
+
++
+*Example*
+
++
+The following examples show that the AES_ENCRYPT function returns the different results based on the different modes. 
+
++
+** The mode aes_192_ecb is in effect.
+
++
+```
+CQD BLOCK_ENCRYPTION_MODE '1';
+
+--- SQL operation complete.
+```
+
++
+```
+SELECT AES_ENCRYPT ('Technical_Writer','Coder') FROM DUAL;
+
+(EXPR)
+--------------------------------
+4}\îî¢Ø­ã^ûh<á¼Æ7õé³ê!cH^Ð
+```
+
++
+** The mode aes_256_ecb is in effect. 
+
++
+```
+CQD BLOCK_ENCRYPTION_MODE '2';
+
+--- SQL operation complete.
+```
+
++
+```
+SELECT AES_ENCRYPT ('Technical_Writer','Coder') FROM DUAL;
+
+(EXPR)
+--------------------------------
+Ô_8ÆðÑ=Uú2g1¥@°Ô3yä­
+
+--- 1 row(s) selected.
+```
+
+[[considerations_for_aes_encrypt]]
+=== Considerations for AES_ENCRYPT 
+
+Either argument cannot be NULL. 
+
+[[examples_of_aes_encrypt]]
+=== Examples of AES_ENCRYPT
+
+* This example encrypts the string _'2019-11-02 00:00:00'_ with the key _'888'_.
+
++
+```
+SELECT AES_ENCRYPT ('2019-11-02 00:00:00', '888') FROM DUAL;
+
+(EXPR)
+--------------------------------
+ºMµ-b(úaP\ìqXqsnTç<`®
+
+--- 1 row(s) selected.
+```
+
+* This example encrypts the string _'Oct 9, 2013 was my first day in Muenster'_ with the key _'Germany-10:06'_ and inserts the encrypted string into the table _t1_.
+
```

Comment:
Probably need an AES_DECRYPT function (which exists in the product) to return the encrypted data to its original form.  Do you also plan to document AES_DECRYPT?

Reply:
Before documenting the AES_DECRYPT function, I've tested the AES_DECRYPT function and encountered a problem as below:

The AES_DECRYPT function works: 
```
SELECT AES_DECRYPT(AES_ENCRYPT('aaa','b'),'b') FROM DUAL;

(EXPR)
----------------
aaa             

--- 1 row(s) selected.
```

but fails:

- Example 1
```
SELECT AES_ENCRYPT ('aaa','b') FROM DUAL;

(EXPR)
----------------
"-=@ìÎzíÃ¸Ä(

--- 1 row(s) selected.
```

```
SELECT AES_DECRYPT ('"-=@ìÎzíÃ¸Ä(','b') FROM DUAL;

*** ERROR[8957] OpenSSL return error int the call to function AES_DECRYPT FUNCTION [2019-04-08 07:00:21]
```

```
SQL>SELECT AES_DECRYPT ("-=@ìÎzíÃ¸Ä(,'b') FROM DUAL;

*** ERROR[15005] Unmatched quote in input (unterminated string): 
SELECT AES_DECRYPT ("-=@ìÎzíÃ¸Ä(,'b') FROM DUAL; [2019-04-08 07:09:44]
```
- Example 2
```
SELECT * FROM t1;

C1
---------------------------------------------------------------------------------------------------
a                                                                                                  
b                                                                                                  
c                                                                                                  

--- 3 row(s) selected.
```

```
SELECT c1, AES_DECRYPT (c1, 'd') FROM t1; 

*** ERROR[8957] OpenSSL return error int the call to function AES_DECRYPT FUNCTION [2019-04-08 07:49:57]
```

Is there something wrong with my test case?

Reply:
I am not familiar with the AES_DECRYPT internals but it looks like the data being returned causes issues with the terminal character set.  It works fine when using Hex: 

SELECT converttohex (AES_ENCRYPT ('aaa','b')) FROM DUAL -> returns
0D15222D3D8E40ECCE7AEDC3B8C4280D
select aes_decrypt (X'0D15222D3D8E40ECCE7AEDC3B8C4280D', 'b') from dual -> returns
aaa 

However, if I:
create table t1 (a char (100));
insert into t1 select aes_encrypt('aaa', 'b') from dual;
select (aes_decrypt(a, 'b')) from t1
select converttohex(aes_decrypt(a, 'b')) from t1 -> both return:
*** ERROR[8957] OpenSSL returned error during call to function AES_DECRYPT FUNCTION.

but select converttohex(a) from t1 -> returns that same data as above spaced filled.

I looked in SQL regressions and there was no test case where data was inserted encrypted.

Reply:
@robertamarton thanks for your help. 
I'll merge this PR and add details until we have more.